FLT/Deformations/Algebra/InverseLimit/Basic.lean

```diff
@@ -0,0 +1,472 @@
+    NonUnitalCommSemiring (InverseLimit G f) where
+  mul_comm := mul_comm
+
+instance [∀ i, NonAssocSemiring (G i)] [∀ i j h, RingHomClass (T h) (G j) (G i)] :
+    NonAssocSemiring (InverseLimit G f) where
+  one_mul := one_mul
+  mul_one := mul_one
+  natCast_zero := Nat.cast_zero
+  natCast_succ := Nat.cast_succ
+
+-- There is no NonAssocCommSemiring
+
+instance [∀ i, Semiring (G i)] [∀ i j h, RingHomClass (T h) (G j) (G i)] :
+    Semiring (InverseLimit G f) where
+  __ : NonAssocSemiring _ := inferInstance
+  __ : Monoid _ := inferInstance
+
+instance [∀ i, CommSemiring (G i)] [∀ i j h, RingHomClass (T h) (G j) (G i)] :
+    CommSemiring (InverseLimit G f) where
+  mul_comm := mul_comm
+
+instance [∀ i, Ring (G i)] [∀ i j h, RingHomClass (T h) (G j) (G i)] : Ring (InverseLimit G f) where
+  __ : Semiring _ := inferInstance
+  __ : AddCommGroupWithOne _ := inferInstance
+
+instance [∀ i, CommRing (G i)] [∀ i j h, RingHomClass (T h) (G j) (G i)] :
+    CommRing (InverseLimit G f) where
+  mul_comm := mul_comm
+
+section Action
+
+instance [∀ i, Zero (G i)] [∀ i, SMulZeroClass R (G i)]
+    [∀ i j h, ZeroHomClass (T h) (G j) (G i)]
+    [∀ i j h, MulActionHomClass (T h) R (G j) (G i)] :
+    SMulZeroClass R (InverseLimit G f) where
+  smul_zero r := by
+    ext i
+    simp
+
+instance [Zero R] [∀ i, Zero (G i)] [∀ i, SMulWithZero R (G i)]
+    [∀ i j h, MulActionHomClass (T h) R (G j) (G i)]
+    [∀ i j h, ZeroHomClass (T h) (G j) (G i)] :
+    SMulWithZero R (InverseLimit G f) where
+  zero_smul x := by
+    ext i
+    simp
+
+instance [∀ i, AddZeroClass (G i)] [∀ i, DistribSMul R (G i)]
+    [∀ i j h, AddMonoidHomClass (T h) (G j) (G i)]
+    [∀ i j h, MulActionHomClass (T h) R (G j) (G i)] :
+    DistribSMul R (InverseLimit G f) where
+  smul_add r x y := by
+    ext i
+    simp
+
+instance [Monoid R] [∀ i, AddMonoid (G i)] [∀ i, DistribMulAction R (G i)]
+    [∀ i j h, DistribMulActionHomClass (T h) R (G j) (G i)] :
+    DistribMulAction R (InverseLimit G f) :=
+  letI _ i j h : MulActionHomClass (T h) R (G j) (G i) := inferInstance
+  { smul_zero := smul_zero, smul_add := smul_add }
+
+instance [Monoid R] [∀ i, Monoid (G i)] [∀ i, MulDistribMulAction R (G i)]
+    [∀ i j h, MonoidHomClass (T h) (G j) (G i)]
+    [∀ i j h, MulActionHomClass (T h) R (G j) (G i)] :
+    MulDistribMulAction R (InverseLimit G f) where
+  smul_mul r x y := by
+    ext i; simp
+  smul_one r := by
+    ext i; simp
+
+instance [Semiring R] [∀ i, AddCommMonoid (G i)] [∀ i, Module R (G i)]
+    [∀ i j h, LinearMapClass (T h) R (G j) (G i)] :
+    Module R (InverseLimit G f) :=
+  letI _ i j h : DistribMulActionHomClass (T h) R (G j) (G i) := inferInstance
+  {
+    add_smul r s x := by ext i; simp [add_smul],
+    zero_smul x := by ext i; simp
+  }
+
+end Action
+
+section ToComponent
+
+variable (G f) in
+def toComponent (i : ι) : InverseLimit G f → G i := fun (z : InverseLimit G f) ↦ z.1 i
```

Comment:
My guess is that CI is now complaining that you didn't give this a docstring. When you do, can you please also mention at least one of the variants in the `toComponent` docstring (and hint at all of them if you don't mention them)? And can you face calling them the mathlib names i.e. `toComponentMonoidHom` etc?